File: typescript/advanced/03-batch-operations.ts
/**
 * Advanced API Example 03: Batch Operations
 *
 * Demonstrates processing multiple transactions in parallel.
 * Useful for high-throughput agent scenarios.
 *
 * This example demonstrates:
 * - Creating multiple transactions concurrently
 * - Parallel state transitions
 * - Batch status checking
 * - Error handling in batch operations
 *
 * Run: npx ts-node advanced/03-batch-operations.ts
 */

import 'dotenv/config';
import { ACTPClient } from '@agirails/sdk';
import { log, logSection, formatState } from '../src/utils/helpers';

async function main() {
  logSection('AGIRAILS Advanced API - Batch Operations');

  const requesterAddress = '0x1111111111111111111111111111111111111111';

  log('🔧', 'Creating client...');
  const client = await ACTPClient.create({
    mode: 'mock',
    requesterAddress,
  });

  // Mint tokens for batch operations
  await client.mintTokens(requesterAddress, '1000000000'); // 1000 USDC
  log('✅', 'Client ready with 1000 USDC');

  // =====================================================
  // Batch Create Transactions
  // =====================================================
  logSection('Step 1: Batch Create Transactions');

  const providers = [
    '0x2222222222222222222222222222222222222222',
    '0x3333333333333333333333333333333333333333',
    '0x4444444444444444444444444444444444444444',
    '0x5555555555555555555555555555555555555555',
    '0x6666666666666666666666666666666666666666',
  ];

  log('📝', `Creating ${providers.length} transactions in parallel...`);
  const startCreate = Date.now();

  const createPromises = providers.map((provider, index) =>
    client.intermediate.createTransaction({
      provider,
      amount: String(10 + index * 5), // $10, $15, $20, $25, $30
      deadline: '+2h',
      disputeWindow: 3600, // 1 hour (minimum)
    })
  );

  const txIds = await Promise.all(createPromises);
  const createTime = Date.now() - startCreate;

  console.log(`   Created ${txIds.length} transactions in ${createTime}ms`);
  txIds.forEach((id, i) => {
    console.log(`   ${i + 1}. ${id.substring(0, 16)}... → Provider ${i + 1}`);
  });

  // =====================================================
  // Batch Link Escrow
  // =====================================================
  logSection('Step 2: Batch Link Escrow');

  log('💳', 'Linking escrow for all transactions...');
  const startLink = Date.now();

  const linkPromises = txIds.map((txId) => client.intermediate.linkEscrow(txId));

  await Promise.all(linkPromises);
  const linkTime = Date.now() - startLink;

  console.log(`   Linked ${txIds.length} escrows in ${linkTime}ms`);

  // =====================================================
  // Batch Status Check
  // =====================================================
  logSection('Step 3: Batch Status Check');

  log('🔍', 'Checking all transaction statuses...');

  const statusPromises = txIds.map((txId) => client.intermediate.getTransaction(txId));
  const transactions = await Promise.all(statusPromises);

  console.log('   Status report:');
  transactions.forEach((tx, i) => {
    console.log(`   ${i + 1}. ${formatState(tx!.state)} - $${Number(tx!.amount) / 1_000_000} USDC`);
  });

  // Count by state
  const stateCounts: Record<string, number> = {};
  transactions.forEach((tx) => {
    stateCounts[tx!.state] = (stateCounts[tx!.state] || 0) + 1;
  });
  console.log('\n   State distribution:', stateCounts);

  // =====================================================
  // Batch State Transitions
  // =====================================================
  logSection('Step 4: Batch State Transitions');

  log('🔨', 'Transitioning all to IN_PROGRESS...');
  const startTransition = Date.now();

  const progressPromises = txIds.map((txId) =>
    client.intermediate.transitionState(txId, 'IN_PROGRESS')
  );

  await Promise.all(progressPromises);
  const transitionTime = Date.now() - startTransition;

  console.log(`   Transitioned ${txIds.length} transactions in ${transitionTime}ms`);

  log('📦', 'Transitioning all to DELIVERED...');

  const deliverPromises = txIds.map((txId) =>
    client.intermediate.transitionState(txId, 'DELIVERED')
  );

  await Promise.all(deliverPromises);
  console.log(`   All transactions delivered`);

  // =====================================================
  // Error Handling in Batch
  // =====================================================
  logSection('Step 5: Error Handling');

  log('⚠️', 'Demonstrating batch error handling...');

  // Try to create transactions with one invalid provider
  const mixedProviders = [
    '0x7777777777777777777777777777777777777777', // Valid
    '0x8888888888888888888888888888888888888888', // Valid
    'invalid-address', // Invalid - will fail
  ];

  console.log('   Creating batch with one invalid address...');

  const results = await Promise.allSettled(
    mixedProviders.map((provider) =>
      client.intermediate.createTransaction({
        provider,
        amount: '5',
        deadline: '+1h',
        disputeWindow: 3600, // 1 hour minimum
      })
    )
  );

  let successCount = 0;
  let failCount = 0;

  results.forEach((result, i) => {
    if (result.status === 'fulfilled') {
      successCount++;
      console.log(`   ✅ Transaction ${i + 1}: ${result.value.substring(0, 16)}...`);
    } else {
      failCount++;
      console.log(`   ❌ Transaction ${i + 1}: ${result.reason.message}`);
    }
  });

  console.log(`\n   Results: ${successCount} succeeded, ${failCount} failed`);

  // =====================================================
  // Performance Summary
  // =====================================================
  logSection('Performance Summary');

  console.log('Batch operation timings:');
  console.log(`  Create ${txIds.length} transactions: ${createTime}ms`);
  console.log(`  Link ${txIds.length} escrows: ${linkTime}ms`);
  console.log(`  Transition ${txIds.length} states: ${transitionTime}ms`);
  console.log('');
  console.log('Tips for production:');
  console.log('  - Use Promise.allSettled() for error resilience');
  console.log('  - Batch size 10-50 for optimal performance');
  console.log('  - Implement retry logic for failed operations');
  console.log('  - Monitor gas costs when batching on-chain');

  // Final status
  console.log(`\nTotal transactions created: ${txIds.length}`);

  log('🎉', 'Batch operations demo complete!');
}

main().catch(console.error);

File: python/advanced/03_batch_operations.py
#!/usr/bin/env python3
"""
Advanced API Example 03: Batch Operations

Demonstrates parallel transaction creation and state transitions.
Useful for high-throughput applications.

This example demonstrates:
- Parallel transaction creation
- Batch state transitions
- Performance metrics
- Error handling in batch operations

Run: python advanced/03_batch_operations.py
"""

import asyncio
import time
import sys
from pathlib import Path
from typing import List

sys.path.insert(0, str(Path(__file__).parent.parent))

from src.utils.helpers import clear_mock_state, log, log_section, format_usdc


async def main() -> None:
    log_section("AGIRAILS Advanced API - Batch Operations")

    clear_mock_state()

    from agirails import ACTPClient

    requester_address = "0x1111111111111111111111111111111111111111"
    provider_address = "0x2222222222222222222222222222222222222222"

    log("🔧", "Creating client...")

    client = await ACTPClient.create(
        mode="mock",
        requester_address=requester_address,
    )

    # Mint tokens for batch operations
    await client.mint_tokens(requester_address, 1_000_000_000)  # 1000 USDC
    log("✅", "Client ready with 1000 USDC")

    # =====================================================
    # Part 1: Sequential vs Parallel Transaction Creation
    # =====================================================
    log_section("Part 1: Sequential vs Parallel Creation")

    BATCH_SIZE = 5

    # Sequential creation
    log("🐢", f"Creating {BATCH_SIZE} transactions sequentially...")
    start = time.time()

    sequential_tx_ids: List[str] = []
    for i in range(BATCH_SIZE):
        tx_id = await client.standard.create_transaction({
            "provider": provider_address,
            "amount": 10,  # $10 each
            "deadline": "+1h",
            "dispute_window": 3600,
        })
        sequential_tx_ids.append(tx_id)

    sequential_time = time.time() - start
    print(f"   Created {len(sequential_tx_ids)} transactions")
    print(f"   Time: {sequential_time:.2f}s")
    print(f"   Rate: {BATCH_SIZE / sequential_time:.1f} tx/s")
    print()

    # Parallel creation
    log("🚀", f"Creating {BATCH_SIZE} transactions in parallel...")
    start = time.time()

    async def create_tx(i: int) -> str:
        return await client.standard.create_transaction({
            "provider": provider_address,
            "amount": 10,
            "deadline": "+1h",
            "dispute_window": 3600,
        })

    parallel_tx_ids = await asyncio.gather(*[create_tx(i) for i in range(BATCH_SIZE)])

    parallel_time = time.time() - start
    print(f"   Created {len(parallel_tx_ids)} transactions")
    print(f"   Time: {parallel_time:.2f}s")
    print(f"   Rate: {BATCH_SIZE / parallel_time:.1f} tx/s")
    print()

    speedup = sequential_time / parallel_time if parallel_time > 0 else 0
    print(f"   Speedup: {speedup:.1f}x faster with parallel execution")

    # =====================================================
    # Part 2: Batch State Transitions
    # =====================================================
    log_section("Part 2: Batch State Transitions")

    # Link escrow for all parallel transactions
    log("💳", f"Linking escrow for {BATCH_SIZE} transactions...")
    start = time.time()

    async def link_escrow(tx_id: str) -> str:
        await client.standard.link_escrow(tx_id)
        return tx_id

    await asyncio.gather(*[link_escrow(tx_id) for tx_id in parallel_tx_ids])
    print(f"   Completed in {time.time() - start:.2f}s")

    # Transition all to DELIVERED
    log("📦", f"Transitioning {BATCH_SIZE} transactions to DELIVERED...")
    start = time.time()

    async def transition_to_delivered(tx_id: str) -> str:
        await client.standard.transition_state(tx_id, "DELIVERED")
        return tx_id

    await asyncio.gather(*[transition_to_delivered(tx_id) for tx_id in parallel_tx_ids])
    print(f"   Completed in {time.time() - start:.2f}s")

    # =====================================================
    # Part 3: Error Handling with allSettled Pattern
    # =====================================================
    log_section("Part 3: Error Handling")

    log("⚠️", "Demonstrating error handling with mixed results...")

    # Mix of valid and invalid transaction IDs
    test_tx_ids = parallel_tx_ids[:3] + [
        "0x" + "0" * 64,  # Invalid TX ID
        "0x" + "1" * 64,  # Another invalid TX ID
    ]

    async def get_tx_safe(tx_id: str) -> dict:
        """Get transaction with error handling."""
        try:
            tx = await client.standard.get_transaction(tx_id)
            return {"status": "success", "tx_id": tx_id, "tx": tx}
        except Exception as e:
            return {"status": "error", "tx_id": tx_id, "error": str(e)}

    results = await asyncio.gather(*[get_tx_safe(tx_id) for tx_id in test_tx_ids])

    success_count = sum(1 for r in results if r["status"] == "success")
    error_count = sum(1 for r in results if r["status"] == "error")

    print(f"   Total: {len(results)}")
    print(f"   Success: {success_count}")
    print(f"   Errors: {error_count}")
    print()

    # Show errors
    for r in results:
        if r["status"] == "error":
            print(f"   ❌ {r['tx_id'][:16]}... - {r['error'][:40]}...")

    # =====================================================
    # Part 4: Performance Metrics
    # =====================================================
    log_section("Part 4: Performance Metrics")

    # Measure total throughput
    total_txs = len(sequential_tx_ids) + len(parallel_tx_ids)
    total_amount = total_txs * 10_000_000  # 10 USDC each

    print("Batch Operation Summary:")
    print(f"  Total transactions: {total_txs}")
    print(f"  Total amount: {format_usdc(total_amount)}")
    print()
    print("Performance:")
    print(f"  Sequential: {BATCH_SIZE / sequential_time:.1f} tx/s")
    print(f"  Parallel: {BATCH_SIZE / parallel_time:.1f} tx/s")
    print(f"  Improvement: {speedup:.1f}x")
    print()
    print("Recommendations:")
    print("  - Use parallel operations for batch processing")
    print("  - Group transactions by provider for efficiency")
    print("  - Use allSettled pattern for resilience")
    print("  - Monitor gas costs in mainnet (Base L2 is cheap)")

    # =====================================================
    # Summary
    # =====================================================
    log_section("Summary")

    print(f"Total transactions created: {total_txs}")
    print(f"  - Sequential batch: {len(sequential_tx_ids)}")
    print(f"  - Parallel batch: {len(parallel_tx_ids)}")

    # Get final balance
    final_balance = await client.get_balance(requester_address)
    print(f"\nFinal balance: {format_usdc(final_balance)}")

    log("🎉", "Batch operations demo complete!")


if __name__ == "__main__":
    asyncio.run(main())
